File: 84. Largest Rectangle in Histogram/main.py
from typing import List

class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        new_stack: List[tuple[int,int]] = []
        result = 0

        for i in range(len(heights)):
            start = i

            while new_stack and new_stack[-1][1] > heights[i]:
                idx, h = new_stack.pop()
                result = max(result, h * (i - idx))
                start = idx

            new_stack.append((start, heights[i]))

        while new_stack:
            temp = new_stack.pop()
            result = max(result, temp[1] * (len(heights) - temp[0]))

        return result

File: 84. Largest Rectangle in Histogram/main.js
/**
 * @param {number[]} heights
 * @return {number}
 */
const largestRectangleArea = function(heights) {
  const stack = []; // Elements: [index, height]
  let maxArea = 0;

  for (let i = 0; i < heights.length; i++) {
    let start = i;
    while (stack.length > 0 && stack[stack.length - 1][1] > heights[i]) {
      const [idx, h] = stack.pop();
      maxArea = Math.max(maxArea, h * (i - idx));
      start = idx;
    }
    stack.push([start, heights[i]]);
  }

  while (stack.length > 0) {
    const [idx, h] = stack.pop();
    maxArea = Math.max(maxArea, h * (heights.length - idx));
  }

  return maxArea;
};
